packages/next/src/build/webpack/loaders/app-dir/metadata.ts

```diff
@@ -0,0 +1,128 @@
+import path from 'path'
+import { stringify } from 'querystring'
+
+const METADATA_TYPE = 'metadata'
+
+export const METADATA_IMAGE_RESOURCE_QUERY = '?__next_metadata'
+
+const staticAssetIconsImage = {
+  icon: {
+    filename: 'icon',
+    extensions: ['ico', 'jpg', 'png', 'svg'],
+  },
+  apple: {
+    filename: 'apple-touch-icon',
+    extensions: ['jpg', 'png', 'svg'],
+  },
+}
+
+// Produce all compositions with filename (icon, apple-touch-icon, etc.) with extensions (png, jpg, etc.)
+async function enumMetadataFiles(
+  dir: string,
+  filename: string,
+  extensions: string[],
+  {
+    resolvePath,
+    addDependency,
+    addMissingDependency,
+  }: {
+    resolvePath: (pathname: string) => Promise<string>
+    addDependency: (dep: string) => any
+    addMissingDependency: (dep: string) => any
+  }
+) {
+  const collectedFiles: string[] = []
+  // Possible filename without extension could: icon, icon0, ..., icon9
+  const possibleFileNames = [filename].concat(
+    Array(10)
+      .fill(0)
+      .map((_, index) => filename + index)
+  )
+  for (const name of possibleFileNames) {
+    for (const ext of extensions) {
+      const pathname = path.join(dir, `${name}.${ext}`)
+      try {
+        const resolved = await resolvePath(pathname)
+        addDependency(resolved)
+
+        collectedFiles.push(resolved)
+      } catch (err: any) {
+        if (!err.message.includes("Can't resolve")) {
+          throw err
+        }
+        addMissingDependency(pathname)
+      }
+    }
+  }
```

Comment:
This is an interesting work around, do we have any idea to make it more general in the future?

Reply:
so far webpack can't watch the infinite set of files, so we changed to 0 - 9, the icon list won't be super large, should be enough, otherwise can fallback to the static exports in page/layout